---
title: "demonstrate_bias_breast_cancer"
output: html_document
date: '2023-01-25'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

#downloaded from https://www.10xgenomics.com/resources/datasets/human-breast-cancer-visium-fresh-frozen-whole-transcriptome-1-standard
#wget https://cf.10xgenomics.com/samples/spatial-exp/1.3.0/Visium_Human_Breast_Cancer/Visium_Human_Breast_Cancer_raw_feature_bc_matrix.tar.gz

library(SpatialExperiment)
library(ggspavis)
library(nnSVG)
library(scran)

(spe <- read10xVisium("/users/kshah/mean_var_project/breast_cancer_data/outs",
  type = "sparse",   # use sparse (not HDF5) format
  data = "raw",     
  images = "lowres", # specify which image(s) to include
  load = TRUE))      # specify whether or not to load image(s)

#plotSpots(spe)

#removes 94 spots not in tissue
spe <- spe[, colData(spe)$in_tissue == 1]

#filter low expressed genes
n_umis <- 100
ix_low_genes <- rowSums(counts(spe)) < n_umis
table(ix_low_genes)
spe <- spe[!ix_low_genes, ]
dim(spe)

#nnSVG needs logcounts
spe <- logNormCounts(spe)

# set seed for reproducibility
set.seed(123)
#run nnSVG
spe <- nnSVG(spe)

# show results
rowData(spe)

# Save an object to a file
saveRDS(spe, file = "spe_breast_cancer_nnSVG.rds")

```


```{r plotting}

library(SpatialExperiment)
library(here)
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(ggrepel)
library(viridis)
library(ggpubr)

spe <- readRDS(file = "mean_var_project/spe_breast_cancer_nnSVG.rds")

spe_df <- rowData(spe)

df_nnSVG <- 
  as.data.frame(spe_df)

df_effect <- 
  as.data.frame(df_nnSVG) %>% 
  mutate(l = 1 / phi) %>% 
  filter(rank <= 1000)

# variance vs. mean
var <- ggplot(df_effect, 
       aes(x = mean, y = var, color = LR_stat)) + 
  geom_point(size = 2) + 
  scale_color_viridis(trans = "log10") +
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "mean logcounts", 
       y = "variance", 
       color = "LR statistic") + 
  ggtitle("nnSVG: human breast") + 
  theme_bw()

# spatial variance vs. mean
spat_var <- ggplot(df_effect, 
       aes(x = mean, y = sigma.sq, color = LR_stat)) + 
  geom_point(size = 2) + 
  scale_color_viridis(trans = "log10") + 
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "mean logcounts", 
       y = "spatial variance (sigma^2)", 
       color = "LR statistic") + 
  ggtitle("nnSVG: human HPC") + 
  theme_bw()

# nonspatial variance vs. mean
nonspat_var <- ggplot(df_effect, 
       aes(x = mean, y = tau.sq, color = LR_stat)) + 
  geom_point(size = 2) + 
  scale_color_viridis(trans = "log10") + 
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "mean logcounts", 
       y = "nonspatial variance (tau^2)", 
       color = "LR statistic") + 
  ggtitle("nnSVG: human HPC") + 
  theme_bw()

# proportion spatial variance vs. mean
prop <- ggplot(df_effect, 
       aes(x = mean, y = prop_sv, color = LR_stat)) + 
  geom_point(size = 2) + 
  scale_color_viridis(trans = "log10") + 
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "mean logcounts", 
       y = "proportion spatial variance", 
       color = "LR statistic") + 
  ggtitle("nnSVG: human HPC") + 
  theme_bw()

figure <- ggarrange(var, spat_var, nonspat_var, prop,
          labels = c("A", "B", "C", "D"),
          ncol = 2, nrow = 2,
          common.legend = TRUE, legend = "right")

title_text <- "human breast"
annotate_figure(figure, 
                top = text_grob(title_text, color = "black", face = "bold", size = 14))

```




